File: app/src/main/java/com/bapplications/maplemobile/gameplay/textures/Frame.java
package com.bapplications.maplemobile.gameplay.textures;

import android.util.Pair;

import com.bapplications.maplemobile.opengl.utils.Point;
import com.bapplications.maplemobile.opengl.utils.Rectangle;
import com.bapplications.maplemobile.pkgnx.NXNode;
import com.bapplications.maplemobile.pkgnx.nodes.NXLongNode;

public class Frame extends Texture {
    private Pair<Integer, Integer> scales;
    private Pair<Integer, Integer> opacities;
    private short delay;
    private Point head;
    private Rectangle bounds;

    public Frame(NXNode src) {
        super(src);
        bounds = new Rectangle(src);
        head = new Point(src.getChild("head"));
        head.flipY();
        try {
            delay = ((Long)(src.getChild("delay").get())).shortValue();
        } catch (NullPointerException e){
            delay = 0;
        }

        if (delay == 0)
            delay = 100;

        NXNode a0 = src.getChild("a0");
        NXNode a1 = src.getChild("a1");
        boolean hasa0 = a0 != null && a0.get() instanceof NXLongNode;
        boolean hasa1 = a1 != null && a1.get() instanceof NXLongNode;

        if (hasa0 && hasa1)
        {
            opacities = new Pair(a0.get(), a1.get());
        }
        else if (hasa0)
        {
            byte a0v = (byte)a0.get();
            opacities = new Pair(a0v, 255 - a0v);
        }
        else if (hasa1)
        {
            byte a1v = (byte)a1.get();
            opacities = new Pair(255 - a1v, a1v);
        }
        else
        {
            opacities = new Pair(255, 255);
        }


        NXNode z0 = src.getChild("z0");
        NXNode z1 = src.getChild("z1");
        boolean hasz0 = z0 != null && z0.get() instanceof NXLongNode;
        boolean hasz1 = z1 != null && z1.get() instanceof NXLongNode;

        if (hasz0 && hasz1)
        {
            scales = new Pair(z0.get(), z1.get());
        }
        else if (hasz0)
        {
            byte z0v = (byte)z0.get();
            scales = new Pair(z0v, 100 - z0v);
        }
        else if (hasz1)
        {
            byte z1v = (byte)z1.get();
            scales = new Pair(100 - z1v, z1v);
        }
        else
        {
            scales = new Pair(100, 100);
        }
    }

    public Frame() {
        super();
        delay = 0;
        opacities = new Pair(0, 0);
        scales = new Pair(0, 0);
    }


    public int startOpacity()
    {
        return opacities.first;
    }

    public int startScale()
    {
        return scales.first;
    }

    public short getDelay() {
        return delay;
    }

    public float opcstep(int timestep) {
        return timestep * (float)(opacities.second - opacities.first) / delay;
    }

    public float scalestep(int timestep) {
        return timestep * (float)(scales.second - scales.first) / delay;
    }

    public void setDelay(short delay) {
        this.delay = delay;
    }

    public Point getHead() {
        return head;
    }
}

File: app/src/main/java/com/bapplications/maplemobile/gameplay/textures/Frame.kt
package com.bapplications.maplemobile.gameplay.textures

import android.util.Pair
import com.bapplications.maplemobile.pkgnx.NXNode
import com.bapplications.maplemobile.utils.Point
import com.bapplications.maplemobile.utils.Rectangle

class Frame : Texture {
    lateinit var head: Point
        private set
    var delay: Short = 0
    lateinit var bounds: Rectangle
        private set
    private var scales: Pair<Byte, Byte>? = null
    private var opacities: Pair<Byte, Byte>? = null

    constructor(src: NXNode, recycle: Boolean = true) : super(src, true, recycle) {
        bounds = Rectangle(src)
        head = Point(src.getChild("head"))
        head.flipY()
        delay = try {
            src.getChild<NXNode>("delay").get(0L).toShort()
        } catch (e: ClassCastException) {
            // There are cases where delay is string node and not a numeric node
            src.getChild<NXNode>("delay").get("0").toShort()
        }
        if (delay.toInt() == 0) delay = 100
        val a0 = src.getChild<NXNode>("a0")
        val a1 = src.getChild<NXNode>("a1")
        val hasa0 = !a0.isNotExist
        val hasa1 = !a1.isNotExist
        opacities = if (hasa0 && hasa1) {
            Pair(a0.get(0L).toByte(), a1.get(0L).toByte())
        } else if (hasa0) {
            val a0v = a0.get(0L).toByte()
            Pair(a0v, (255 - a0v).toByte())
        } else if (hasa1) {
            val a1v = a1.get(0L).toByte()
            Pair((255 - a1v).toByte(), a1v)
        } else {
            Pair(255.toByte(), 255.toByte())
        }
        val z0 = src.getChild<NXNode>("z0")
        val z1 = src.getChild<NXNode>("z1")
        val hasz0 = !z0.isNotExist
        val hasz1 = !z1.isNotExist
        if (hasz0 && hasz1) {
            scales = Pair<Byte, Byte>(z0.get(0L).toByte(), z1.get(0L).toByte())
        } else if (hasz0) {
            val z0v = z0.get(0L).toByte()
            scales = Pair<Byte, Byte>(z0v, (100 - z0v).toByte())
        } else if (hasz1) {
            val z1v = z1.get(0L).toByte()
            scales = Pair<Byte, Byte>((100 - z1v).toByte(), z1v)
        } else {
            scales = Pair<Byte, Byte>(100.toByte(), 100.toByte())
        }
    }

    constructor() : super() {
        delay = 0
        opacities = Pair(0.toByte(), 0.toByte())
        scales = Pair(0.toByte(), 0.toByte())
    }

    fun startOpacity(): Byte {
        return opacities!!.first
    }

    fun startScale(): Byte {
        return scales!!.first
    }

    fun opcstep(timestep: Int): Float {
        return timestep * (opacities!!.second - opacities!!.first).toFloat() / delay
    }

    fun scalestep(timestep: Int): Float {
        return timestep * (scales!!.second!! - scales!!.first).toFloat() / delay
    }
}